hcl2template/addrs/plugin.go

```diff
@@ -13,21 +16,58 @@ import (
 
 // Plugin encapsulates a single plugin type.
 type Plugin struct {
-	Hostname  string
-	Namespace string
-	Type      string
+	Source string
 }
 
-func (p Plugin) RealRelativePath() string {
-	return p.Namespace + "/packer-plugin-" + p.Type
+// Parts returns the list of components of the source URL, starting with the
+// host, and ending with the name of the plugin (non-simplified).
+//
+// This will correspond more or less to the filesystem hierarchy where
+// the plugin is installed.
+func (p Plugin) Parts() []string {
+	var retParts []string
+
+	// This should not happen, we're supposed to check that the source isn't
+	// empty when creating the plugin, but let's be careful here and not
+	// misbehave.
+	if p.Source == "" {
+		log.Printf("[ERROR] - Plugin.Parts: the source was empty, this should not happen in normal usage, and is likely a Packer bug.")
+		return retParts
+	}
+
+	dir, part := path.Split(strings.TrimRight(p.Source, "/"))
+	for dir != "" {
+		retParts = append(retParts, part)
+		dir, part = path.Split(strings.TrimRight(dir, "/"))
```

Comment:
So, `dir` is never empty in that loop, since we are purposefully looping until it becomes empty.
That happens when you split the path, as when you reach the top dir, it becomes empty.

Ex:

```
p.Source = "github.com/hashicorp/amazon"
# Loop 1
dir = "github.com/hashicorp"
part = "amazon"
# Loop 2
dir = "github.com"
part = "hashicorp"
# Loop 3
dir = ""
part = "github.com"
# Out of loop

retParts = ["amazon", "hashicorp", "github.com"]
```

Reply:
Ohh, I didn't see it was a `for`, my bad!!